[Console] Improve UX around handling multiple requests

mibragimov opened the pull request:
Fixes https://github.com/elastic/kibana/issues/130982

This PR adds support for highlighting the results for multiple requests in the editor output

<img width="1495" alt="Screen Shot 2022-05-23 at 17 13 04" src="https://user-images.githubusercontent.com/53621505/169816823-0ea541ad-0703-46d9-88b9-fa662e4bf74d.png">

mibragimov commented on src/plugins/console/public/application/models/legacy_core_editor/mode/output_highlight_rules.js:
```diff
@@ -25,7 +25,30 @@ export function OutputJsonHighlightRules() {
+        return 'badge.badge-danger';
+      },
+      // Matches any digits ending with http statusText at the end of string
+      regex: /(\d+\s(OK|Bad Request|Not Found|Continue|Created))$/,
```
Hi @cjcenizal , I just wanted to check with you. Do we need to localize the HTTP status texts above? (OK, Bad request, etc.)

cjcenizal replied:
Yes I think we should localize them.

mibragimov replied:
Thanks for the suggestions! I've made slight adjustments to the regex pattern to match any digits followed by any letters. WDYT? 

```
(\d+\s[\sA-Za-z]+$)
```

This will avoid matching other numbers at the end of a string in the output panel. E.g. 
<img width="1498" alt="Screen Shot 2022-05-24 at 15 45 29" src="https://user-images.githubusercontent.com/53621505/170018716-5dd506a6-a9ce-4556-a1e5-49729bccdb64.png">

cjcenizal commented on src/plugins/console/public/styles/_app.scss:
```diff
@@ -33,6 +33,42 @@
+    vertical-align: middle;
+    cursor: default;
+    max-width: 100%;
+    text-align: left;
```
In dark mode, the text for this badge appears white:

![image](https://user-images.githubusercontent.com/1238659/169879706-85d06bb0-255e-4aa4-9e57-ac9c42257197.png)

We can add `color: black;` to fix this, but this seems inelegant. @elastic/eui-design Any suggestions on how we should incorporate EUI's coloring rules into these styles, to get this badge to match `EuiBadge`?

![image](https://user-images.githubusercontent.com/1238659/169879601-695340ba-35a0-4dec-9b57-8936bc941774.png)

cchaos replied:
Via: https://elastic.github.io/eui/#/utilities/color-functions#light-or-dark-text?themeLanguage=sass

```sass
color: chooseLightOrDarkText($color);
```

cjcenizal replied:
TY!

cjcenizal commented on src/plugins/console/public/application/components/network_request_status_bar/network_request_status_bar.tsx:
```diff
@@ -66,7 +68,8 @@ export const NetworkRequestStatusBar: FunctionComponent<Props> = ({
     );
-  } else if (requestResult) {
+    // If multiple requests are sent, we do not need to render status badges since they are displayed in the editor output panel
+  } else if (requestResult && !isMultipleRequest) {
```
Let's revert this change, so we can continue to render the "time elapsed" information.

![image](https://user-images.githubusercontent.com/1238659/169880585-c48d14c8-fa73-4363-b47b-cca69b95a515.png)

This does introduce friction between the status code reported here and those reported in the results pane if they aren't all the same value. I think the status code in the header should show the first error encountered in the results pane. So if everything returns 200 except one request returns a 404, then I think the status code here should be 404. I think this is beyond the scope of this PR so after we merge this PR, can you create a followup issue to implement this change?

cjcenizal replied:
Thanks for reverting this change. I created https://github.com/elastic/kibana/issues/132939 to track future enhancements about this behavior.

cjcenizal commented on src/plugins/console/public/application/hooks/use_send_current_request/send_request.ts:
```diff
@@ -106,7 +106,17 @@ export function sendRequest(args: RequestArgs): Promise<RequestResult[]> {
 
             if (isMultiRequest) {
-              value = '# ' + req.method + ' ' + req.url + '\n' + value;
+              value =
```
Looks like we're missing tests for this behavior. Could you please add some?

cjcenizal replied:
Also, can we use a template literal here?

```js
value = `# ${req.method} ${req.url} ${response.status} ${response.statusText}\n`;
```

cjcenizal commented on src/plugins/console/public/application/hooks/use_send_current_request/send_request.ts:
```diff
@@ -141,7 +151,8 @@ export function sendRequest(args: RequestArgs): Promise<RequestResult[]> {
 
         if (isMultiRequest) {
-          value = '# ' + req.method + ' ' + req.url + '\n' + value;
+          value =
```
Can we use a template literal here too?

```js
value = `# ${req.method} ${req.url} ${statusCode} ${statusText}\n${value}`;
```

mibragimov commented:
@elasticmachine merge upstream

cjcenizal commented on src/plugins/console/public/application/containers/editor/legacy/console_editor/editor_output.tsx:
```diff
@@ -11,6 +11,8 @@ import { EuiScreenReaderOnly } from '@elastic/eui';
 import React, { useEffect, useRef } from 'react';
 import { convertMapboxVectorTileToJson } from './mapbox_vector_tile';
+// @ts-ignore
+import * as OutputMode from '../../../../models/legacy_core_editor/mode/output';
```
Yes, let's convert it to TS. Thanks!

mibragimov commented:
@elasticmachine merge upstream

mibragimov commented:
Verified functional tests through flaky-test-suite-runner https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/670

alexwizp commented on test/functional/apps/console/_console.ts:
```diff
@@ -122,5 +122,34 @@ export default function ({ getService, getPageObjects }: FtrProviderContext) {
+        await PageObjects.console.enterRequest('\n GET test-index');
+
+        await sendRequests();
+        await retry.waitFor('success badge', () => PageObjects.console.hasSuccessBadge());
```
to be honest, I don’t really like this approach, in case of an error we will try to find this element again and again. I think it would be more correct to determine that we have updated the output of the console and then check the result `retry.waitFor`

mibragimov commented on test/functional/page_objects/console_page.ts:
```diff
@@ -163,4 +163,28 @@ export class ConsolePageObject extends FtrService {
+  public async hasSuccessBadge(): Promise<boolean> {
+    try {
+      const responseEditor = await this.testSubjects.find('response-editor');
+      return Boolean(responseEditor.findByCssSelector('.ace_badge-success'));
```
Right, good point!

alexwizp commented on src/plugins/console/public/application/models/legacy_core_editor/mode/output.ts:
```diff
@@ -10,23 +10,24 @@ import ace from 'brace';
-oop.inherits(Mode, JSONMode);
 
-(function () {
+(function (this: any) {
```
Can we use?
`(function (this: Mode) {`

alexwizp commented on src/plugins/console/public/styles/_app.scss:
```diff
@@ -33,6 +33,47 @@
+  }
+
+  .ace_badge-default {
+    background-color: $euiColorGhost;
```
just interesting why we need `chooseLightOrDarkText` for `color` and don't need it for `background-colo`r

mibragimov replied:
According to EUI, `chooseLightOrDarkText`is used  _to determine whether or not to use light (white) or dark (black) text against the given background color._ See https://elastic.github.io/eui/#/utilities/color-functions#light-or-dark-text

mibragimov commented on src/plugins/console/public/application/models/legacy_core_editor/mode/output_highlight_rules.ts:
```diff
@@ -0,0 +1,57 @@
+import 'brace/mode/json';
+import { addXJsonToRules } from '@kbn/ace';
+
+const JsonHighlightRules = ace.acequire('ace/mode/json_highlight_rules').JsonHighlightRules;
```
This change causes a ts error on line [15](https://github.com/elastic/kibana/pull/132494/files#diff-5ae03c495eb4866d16401e5ab4e822b3113e89802c65b4eac966e476edcd4f06R15). 

```
TS4020: 'extends' clause of exported class 'OutputJsonHighlightRules' has or is using private name 'JsonHighlightRules'.
```

Any suggestion?

mibragimov commented:
Verified functional tests through flaky-test-suite-runner https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/680

cjcenizal commented on src/plugins/console/public/application/models/legacy_core_editor/mode/output_highlight_rules.ts:
```diff
@@ -0,0 +1,57 @@
+        regex: /#(.*?)(?=\d+\s(?:[\sA-Za-z]+)|$)/,
+      },
+      {
+        token(value: string) {
```
Can we change this line to `token: mapStatusCodeToBadge,` and extract this function onto line 15:

```ts
export const mapStatusCodeToBadge = (value: string) => {
  const regExpMatchArray = value.match(/\d+/);
  if (regExpMatchArray) {
    const status = parseInt(regExpMatchArray[0], 10);
    if (status <= 199) {
      return 'badge.badge--default';
    }
    if (status <= 299) {
      return 'badge.badge--success';
    }
    if (status <= 399) {
      return 'badge.badge--primary';
    }
    if (status <= 499) {
      return 'badge.badge--warning';
    }
    return 'badge.badge--danger';
  }
};
```

Then could you create a sibling file `output_highlight_rules.test.ts` and add these tests? This way if someone makes changes to this logic the tests will catch it. Thanks!

```ts
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { mapStatusCodeToBadge } from './output_highlight_rules';

describe('mapStatusCodeToBadge', () => {
  const testCases = [{
    description: 'treats 100 as as default',
    value: '# PUT test-index 100 Continue',
    badge: 'badge.badge--default',
  }, {
    description: 'treats 200 as success',
    value: '# PUT test-index 200 OK',
    badge: 'badge.badge--success',
  }, {
    description: 'treats 301 as primary',
    value: '# PUT test-index 301 Moved Permanently',
    badge: 'badge.badge--primary',
  }, {
    description: 'treats 400 as warning',
    value: '# PUT test-index 404 Not Found',
    badge: 'badge.badge--warning',
  }, {
    description: 'treats 502 as danger',
    value: '# PUT test-index 502 Bad Gateway',
    badge: 'badge.badge--danger',
  }, {
    description: 'treats unexpected numbers as danger',
    value: '# PUT test-index 666 Demonic Invasion',
    badge: 'badge.badge--danger',
  }, {
    description: 'treats no numbers as undefined',
    value: '# PUT test-index',
    badge: undefined,
  }];

  testCases.forEach(({ description, value, badge }) => {
    test(description, () => {
      expect(mapStatusCodeToBadge(value)).toBe(badge);
    });
  });
});
```

mibragimov replied:
Sure, thanks for the suggestion!

cjcenizal reviewed (approved):
@elastic/kibana-design Would you or some folks from EUI please review the styles here? I want to make sure they're correct, especially for dark mode.

@mibragimov People are going to love this enhancement! Thank you so much for making this change. I had one suggestion to add some more tests, other than that this code LGTM.

Did a quick test locally and things look good there too:

![image](https://user-images.githubusercontent.com/1238659/170550021-6725485d-8883-4f3c-a87a-dc8c530f198f.png)

mibragimov commented:
@elasticmachine merge upstream

mibragimov commented:
@elasticmachine merge upstream

kibanamachine commented:
expected head sha didn’t match current head ref.

MichaelMarcialis commented on src/plugins/console/public/styles/_app.scss:
```diff
@@ -33,6 +33,47 @@
+    vertical-align: middle;
+    cursor: default;
+    max-width: 100%;
+    text-align: left;
```
I believe this style can be removed, as its left text alignment is already being inherited and applied.

```suggestion
```

MichaelMarcialis reviewed (commented):
I've left a few style comments below for your review. Thanks!

cjcenizal commented on src/plugins/console/public/styles/_app.scss:
```diff
@@ -33,6 +33,47 @@
+    text-align: left;
+
+    &--success {
+      background-color: $euiColorSuccess;
+      color: chooseLightOrDarkText($euiColorSuccess);
```
@MichaelMarcialis For our own edification, can you teach us about the role of `$euiColorVis0_behindText` and similar colors, and what makes it better than `$euiColorSuccess` here?

MichaelMarcialis replied:
Good question, @cjcenizal! For the purposes of this PR, my reason for the suggested color changes were purely to make these ace badges consistent with the colors that our `EuiBadge` components use.

That said, I imagine the reason the EUI team ultimately chose to use the behind-text visualization colors for the `EuiBadge` component has to do with accessibility and the contrast ratio, as this component houses fairly small sized text. The contrast with the behind-text visualization colors appears to be higher and receives a better WCAG rating. Unlike `EuiBadge`, I assume that components with larger text sizes (such as `EuiButton`) can afford be a bit less strict with the color contrast, which is why they can use colors like `$euiColorSuccess`.

CCing @cchaos here to keep me honest and double-check my assumptions.

cchaos replied:
For full context on why we used visualization colors for badges, you can read through this PR: https://github.com/elastic/eui/pull/2455

cjcenizal replied:
Thank you both! This helps me out a ton.

mibragimov commented:
@elasticmachine merge upstream

mibragimov commented:
@MichaelMarcialis thanks for the review! I've addressed your feedback with https://github.com/elastic/kibana/pull/132494/commits/ad57319ef66412f6cc8bbf129f112cb1fb2a6912, do you mind having another look?

MichaelMarcialis reviewed (approved):
Thanks for making those changes @mibragimov. I only have one additional thought. Would it be worth changing the font family of `.ace_badge` to `$euiFontFamily` in this case? Currently, the `.ace_badge`s are inheriting the editor font family, which differentiates itself from the standard EUI badges. Assuming these badges only ever appear at the end of editor lines, I think this might look nicer/more consistent.

Otherwise, this looks good from my perspective. Approving now so that I don't hold you up further. Thanks.

mibragimov commented:
@elasticmachine merge upstream
